src/inmanta/util/collections.py

```diff
@@ -0,0 +1,115 @@
+"""
+    Copyright 2024 Inmanta
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+
+from collections.abc import Hashable, Iterator, Mapping, MutableMapping, Set
+from typing import Generic, Optional, TypeVar
+
+P = TypeVar("P", bound=Hashable)
+S = TypeVar("S", bound=Hashable)
+# type vars not bound to the class
+K = TypeVar("K", bound=Hashable)
+V = TypeVar("V", bound=Hashable)
```

Comment:
I found it hard to come up with meaningful names here. I changed it to the conventional `S` and `T` now. I'm not sure comments would be helpful, since the names really don't have much meaning.

Reply:
(If you get notified about these comments, no need to look at them now, I'll make sure to link here from the new PR.)